### PR TITLE
balloon_hotplug:reboot/shutdown guest after plug/unplug balloon device

### PR DESCRIPTION
**Pull request by lijinlijin**
add optional steps to support reboot/shutdown guest after hotplug
or hot-unplug balloon devices

Signed-off-by: Li Jin <lijin@redhat.com>

ID:1456107,1456149,1456150,1456108

### CODE REVIEW

**Comment by lijinlijin** [comment]
Windows 2016 guest:
 (1/5) smp_8.4096m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.balloon_hotplug: PASS (28453.44 s)
 (2/5) smp_8.4096m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.balloon_hotplug.with_reboot: PASS (800.39 s)
 (3/5) smp_8.4096m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.balloon_hotplug.with_system_reset: PASS (798.44 s)
 (4/5) smp_8.4096m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.balloon_hotplug.with_shutdown_after_unplug: PASS (358.06 s)
 (5/5) smp_8.4096m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.balloon_hotplug.with_shutdown_after_plug: PASS (350.22 s)

RHEL7.4 guest:
(1/4) smp_8.4096m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.balloon_hotplug.with_reboot: PASS (717.43 s)
 (2/4) smp_8.4096m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.balloon_hotplug.with_system_reset: PASS (368.54 s)
 (3/4) smp_8.4096m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.balloon_hotplug.with_shutdown_after_unplug: PASS (161.20 s)
 (4/4) smp_8.4096m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.balloon_hotplug.with_shutdown_after_plug: PASS (151.08 s)

File: qemu/tests/balloon_hotplug.py [hunk 1]
```diff
@@ -51,6 +69,13 @@ def run(test, params, env):
                               logging.info)
         balloon_check.run(test, params, env)
 
+        if pm_test_after_plug:
+            run_pm_test(pm_test_after_plug, "hot-plug")
+            if vm.is_alive():
+                balloon_check.run(test, params, env)
+            else:
```
Comment (hereischen) [inline qemu/tests/balloon_hotplug.py]: What do you want to achieve here?
if `vm` is not alive, do nothing?

File: qemu/tests/balloon_hotplug.py [hunk 2]
```diff
@@ -51,6 +69,13 @@ def run(test, params, env):
...
+                return
```
Comment (xiagao) [inline qemu/tests/balloon_hotplug.py]: Should use continue here, if pm_test_after_plug=shutdown, the function will quit, so only run 1 loop

File: qemu/tests/cfg/balloon_hotplug.cfg [hunk 3]
```diff
@@ -14,3 +14,25 @@
     test_tags = "evict enlarge"
     balloon_type_evict = evict
     balloon_type_enlarge = enlarge
+    variants:
+        - @default:
+        - with_reboot:
```
Comment (hereischen) [inline qemu/tests/cfg/balloon_hotplug.cfg]: I think this part should be rearranged, `repeats` and `shell` repeat too many times

File: qemu/tests/balloon_hotplug.py (hunk 2 above)
Comment (hereischen) [reply qemu/tests/balloon_hotplug.py]: Probably just remove `else` part , if it won't do anything

File: qemu/tests/balloon_hotplug.py (hunk 1 above)
Comment (lijinlijin) [reply qemu/tests/balloon_hotplug.py]: pm_test_after_plug mainly include reboot and shutdown.
We need to check if device works well after guest reboot,but it's not necessary if guest is shutdown.
So I will check if guest is alive, if no pm_test_after_plug is shutdown,skip followed tests.
I considered use if pm_test_after_plug == "shutdown": return.But shuang suggest it's not good to filter out a specific case in .py file.

File: qemu/tests/balloon_hotplug.py (hunk 2 above)
Comment (lijinlijin) [reply qemu/tests/balloon_hotplug.py]: Yes,this is what I want.
If guest shut down,no need to execute followed tests.

File: qemu/tests/balloon_hotplug.py [hunk 4]
```diff
@@ -68,3 +93,6 @@ def run(test, params, env):
             raise exceptions.TestFail("Error occurred while hotpluging "
                                       "virtio-pci. Iteration %s, monitor "
                                       "output:%s" % (i, err))
+        else:
```
Comment (xiagao) [inline qemu/tests/balloon_hotplug.py]: else should not be here. 90 line is ok.

File: qemu/tests/balloon_hotplug.py (hunk 1 above)
Comment (lijinlijin) [reply qemu/tests/balloon_hotplug.py]: I will use continue instead of return

File: qemu/tests/balloon_hotplug.py (hunk 2 above)
Comment (lijinlijin) [reply qemu/tests/balloon_hotplug.py]: If remove else part,then followed unplug test will be executed,but actually guest is already shut down.
So I need to break it when pm_test is shutdown

**Comment by lijinlijin** [comment]
@xiagao @hereischen 
Modified
Thanks

**Comment by lijinlijin** [comment]
@hereischen @xiagao 
Any other comments?

File: qemu/tests/cfg/balloon_hotplug.cfg [hunk 5]
```diff
@@ -14,3 +16,19 @@
     test_tags = "evict enlarge"
     balloon_type_evict = evict
     balloon_type_enlarge = enlarge
```
Comment (xiagao) [inline qemu/tests/cfg/balloon_hotplug.cfg]: Seems the three parameters above are not used. Could you have a check by the way?

File: qemu/tests/cfg/balloon_hotplug.cfg [hunk 6]
```diff
@@ -14,3 +16,19 @@
...
+        - with_system_reset:
+            pm_test_after_unplug = boot
+            pm_test_after_plug = boot
+            reboot_method = system_reset
+        - with_shutdown_after_unplug:
+            pm_test_after_unplug = shutdown
+            check_img = yes
+        - with_shutdown_after_plug:
+            pm_test_after_plug = shutdown
+            check_img = yes
```
Comment (xiagao) [inline qemu/tests/cfg/balloon_hotplug.cfg]: I think it's a bit weired here,with_reboot and with_system_reset test cases include unplug and plug actions, but with_shutdown_after_xx  is seperated to two parts.
Why don't change with_shutdown_after_xx  to with_shutdown to keep consistent with reboot and system_reset.

File: qemu/tests/cfg/balloon_hotplug.cfg (hunk 5 above)
Comment (lijinlijin) [reply qemu/tests/cfg/balloon_hotplug.cfg]: These parameters are used for balloon_check subtest.

File: qemu/tests/cfg/balloon_hotplug.cfg (hunk 6 above)
Comment (lijinlijin) [reply qemu/tests/cfg/balloon_hotplug.cfg]: If use only one parameter "with_shutdown",then guest will shutdown after balloon device hotplug,then job finished.
shutdown guest after balloon device unpluged will never be ran.

Comment (xiagao) [reply qemu/tests/cfg/balloon_hotplug.cfg]: Got it ,thanks.
Another question, pm_test_after_unplug and pm_test_after_unplug can be insteaded of pm_test. Why define two parameters?

**Comment by xiagao** [comment]
Acked-by: Xiaoling Gao <xiagao@redhat.com>

File: qemu/tests/balloon_hotplug.py (hunk 1 above)
Comment (hereischen) [reply qemu/tests/balloon_hotplug.py]: OK. Please make a comment in the code, all it maybe difficult for people to understand

Comment (hereischen) [reply qemu/tests/balloon_hotplug.py]: It seems for `pm_test_after_plug == "shutdown"`,  `balloon_repeats` must be 1

Comment (lijinlijin) [reply qemu/tests/balloon_hotplug.py]: Modified,add a line to explain it.
And yes,if run shutdown subtest,the repeat time must be 1,1 is the default value.
Thank you

File: qemu/tests/balloon_hotplug.py [hunk 7]
```diff
@@ -51,6 +69,15 @@ def run(test, params, env):
...
         balloon_check.run(test, params, env)
 
+        if pm_test_after_plug:
+            run_pm_test(pm_test_after_plug, "hot-plug")
+            # run balloon test after reboot,skip followed test if
+            # pm_test_after_plug is shutdown
+            if vm.is_alive():
+                balloon_check.run(test, params, env)
+            else:
+                break
```
Comment (luckyh) [inline qemu/tests/balloon_hotplug.py]: Should be `return`.

File: qemu/tests/balloon_hotplug.py [hunk 8]
```diff
@@ -62,6 +89,9 @@ def run(test, params, env):
         if not ver_out:
             err += ("\nDevice is still in qtree %ss after unplug:\n%s"
                     % (pause, vm.monitor.info("qtree")))
+        else:
+            if pm_test_after_unplug:
+                run_pm_test(pm_test_after_unplug, "hot-unplug")
```
Comment (luckyh) [inline qemu/tests/balloon_hotplug.py]: Better to move this section to right behind L100, and add the following check.
```
if not vm.is_alive():
    return
```

Comment (lijinlijin) [reply qemu/tests/balloon_hotplug.py]: Modified.
Thanks

**Comment by luckyh** [comment]
Thanks all.

**Pull request merged**